### PR TITLE
[NVIDIA TF] Part 2: Fused matmul op supports cudnn matmul fusion

### PR DESCRIPTION
**Pull request by kaixih**
This PR enables the cudnn matmul fusion backend for supporting the generic matmul fusion patterns. Specifically, this PR focuses on the matmul+bias+gelu_exact pattern. (Note, the matmul+bias+gelu_approximate has already been supported by cublasLt backend. See https://github.com/tensorflow/tensorflow/pull/55966)

Part 1: Stream executor supports cudnn matmul fusion.
Part 2: Fused matmul op supports cudnn matmul fusion. (This one)
Part 3: Grappler graph pass supports matmul+bias+gelu_exact.

Note: This PR depends on the changes of a parent PR (https://github.com/tensorflow/tensorflow/pull/56825). Please see the change diff  [here](https://github.com/kaixih/tensorflow/compare/gelu_exact_upstream_part1...kaixih:tensorflow:gelu_exact_upstream_part2?expand=1).

cc. @nluehr @pjannaty

### CODE REVIEW

**Comment by pjannaty** [comment]
Can we change the base from master to Part1? Once Part1 gets merged we can rebase and change the base back to master.

**Comment by kaixih** [comment]
cc @benbarsdell

**Comment by ezhulenev** [comment]
@timshen91 As you added the `conv_autotuning.proto`, I have a question, how do you feel about logging MatmulProto to it :) Specifically with the concrete addresses for inputs.

**Review by ezhulenev — COMMENTED** [review]
Looks good to me, I'll try to find if someone else wants to take a look at it. Ping me once the Part1 will be submitted, so I can check how it looks after rebase.

**Review by pjannaty — COMMENTED** [review]
Thanks, Kaixi!

LGTM for a quick skim (pending Ben's review). Is it at all possible to add any tests to any of the Parts 1-3?

**Review by benbarsdell — APPROVED** [review]
LGTM

File: tensorflow/core/kernels/matmul_op_fused.cc
Comment (kaixih) [reply tensorflow/core/kernels/matmul_op_fused.cc:310]: I like this. Thx.

File: tensorflow/core/kernels/conv_ops_gpu.h
Comment (kaixih) [reply tensorflow/core/kernels/conv_ops_gpu.h:117]: Done.

**Comment by kaixih** [comment]
Rebased and marked as "Ready to review".

**Comment by gbaned** [comment]
@kaixih Can you please resolve conflicts? Thank you!

**Comment by kaixih** [comment]
The rebase is finished. @ezhulenev and @reedwm to review. Thx.

**Comment by gbaned** [comment]
@kaixih Can you please resolve conflicts? Thank you!

**Comment by kaixih** [comment]
Since the stream executor is being moved out, I have to rebase the PRs more frequently than before. The rebase is done. @ezhulenev and @reedwm to review. Thx.

**Review by ezhulenev — APPROVED** [review]

**Comment by reedwm** [comment]
@kaixih Unfortunately the TensorFlow presubmits disallows any protobuf files to be included in a kernel header file, but you included `tensorflow/core/protobuf/autotuning.pb.h` in the file `tensorflow/core/kernels/conv_ops_gpu.h`. I'm not entierly sure what the reason for this presubmit is, but the description of https://github.com/tensorflow/tensorflow/commit/6ada43366663210beb0159b8c1a67b26ebfe6cb7 mentions this.

Can you remove the include from the header? Understandably, this is hard, since you use protobuf messages in the templated `AutotuneConvImpl` function.

**Comment by reedwm** [comment]
Also, in `MatmulParameters` within `tensorflow/core/util/autotune_maps/conv_parameters.h`, you added a new `LINT.IfChange(conv_parameters_version)` line. But there already is a different `LINT.IfChange(conv_parameters_version)` line earlier in the file. Can you fix, or just remove the line?

**Comment by kaixih** [comment]
@reedwm Thanks for the feedback.

I have made the changes accordingly to move the `AutotuneMatMulImpl` and its dependencies to the `matmul_op_fused.cc` instead of the `conv_ops_gpu.h`.

For the `LINT.IfChange(conv_parameters_version)` issue, I proposed a new `LINT.IfChange(matmul_parameters_version)` and pair the corresponding `LINT.If` and `LINK.Then` correctly.

**Comment by reedwm** [comment]
Why reopen this? This has been merged with https://github.com/tensorflow/tensorflow/commit/23dd6bbe697393b97407829ecc32d88821484406.

**Comment by kaixih** [comment]
Sorry, I reopened it because it showed "closed" instead of "merged". And yes, I just noticed it was actually merged. @reedwm 

I got a bit confused. Closing this one for now.

**Comment by reedwm** [comment]
Hmm I wonder why it wasn't marked as merged.

I think the issue might have been I started manually merging this before you forced push a new commit. So I ended up merging an old commit. In particular, I merged at commit https://github.com/tensorflow/tensorflow/commit/4f9cc30f12eaf9b362a03c552f0a9e0bbbfba4d4, but you later force-pushed commit https://github.com/tensorflow/tensorflow/commit/408d38421e5d11b4417bb282cd1e98f91a4edf61. The "merge" commit https://github.com/tensorflow/tensorflow/commit/23dd6bbe697393b97407829ecc32d88821484406 only has one parent, but the descipription does reference https://github.com/tensorflow/tensorflow/commit/4f9cc30f12eaf9b362a03c552f0a9e0bbbfba4d4 in the `COPYBARA_INTEGRATE_REVIEW` line. Probably copybara was unwilling to put it as a parent when it no longer was part of this PR.

**Comment by kaixih** [comment]
I see. I should have done less of the force-pushing. I did that mainly because I wanted to make the other PRs that depend on this one more easily to be rebased.

**Pull request closed without merging**
